File: Graphs/DFS.py
'''
	Given a starting vertex s, 
	traverse through the graph in adjancency matrix format in 
	depth-first manner.
'''
def stack_DFS(adj_mat, s):
	num_ver = len(adj_mat) #number of vertices
	explore_stack = [s] 
	explored = [False] * num_ver
	explored[s-1] = True
	while len(explore_stack) > 0:
		v = explore_stack.pop()
		if not explored[v-1]:
			explored[v-1] = True
			edges = adj_mat[v-1]
			for i in range(num_ver):
				if edges[i] == 1:
					explore_stack.append(i+1)

def stack_al_DFS(vertices, edges, s):
	num_ver = len(vertices)
	explore_stack = [s]
	explored = [False] * num_ver
	explored[s-1] = True
	while len(explore_stack) > 0:
		v = explore_stack.pop()
		if not explored[v-1]:
			explored[v-1] = True
			v_edges = edges[vertices[v-1]:vertices[v]]
			for vertex in v_edges:
				explore_stack.append(vertex)

def topo_sort(vertices, edges):
	num_ver = len(vertices)
	global curlabel = num_ver
	global explored = [False] * num_ver
	global labels = [0] * num_ver
	for v in range(num_ver):
		if not explored[v-1]: 
			DFS_Topo(vertices,edges, e)

def DFS_Topo(vertices, edges, v):
	v_edges = edges[vertices[v-1]:vertices[v]]
	explored[v-1] = True
	for vertex in v_edges:
		if not explored[vertex-1]:
			DFS_Topo(vertices, edges, vertex)
	labels[v-1] = curlabel
	curlabel -= 1

File: Graphs/DFS.cpp
#include <vector>
#include <limits>
#include <stack>

void stack_al_DFS(std::vector<int> vertices, std::vector<int> edges, int s){
	int num_ver = vertices.size();
	std::stack<int> to_be_explored;
	std::vector<bool> explored;
	explored[s-1] = true;
	while(!to_be_explored.empty()){
		int v = to_be_explored.top();
		to_be_explored.pop();
		if(!explored[v-1]){
			explored[v-1] = true;
			std::vector<int> v_edges;
			copy(vertices[v-1],vertices[v],v_edges);
			for (int vertex : v_edges){
				to_be_explored.push(vertex);
			}
		}
	}
}

int curlabel;
std::vector<bool> explored;
std::vector<int> labels;

void dfs_topo(std::vector<int> vertices, std::vector<int> edges, int v){
	int num_ver = vertices.size();
	explored[v-1] = true;
	std::vector<int> v_edges;
	copy(edges[vertices[v-1]],edges[vertices[v]],v_edges);
	for (int vertex : v_edges){
		if (!explored[vertex-1]){
			dfs_topo(vertices, edges, v);
		}
	}
	labels[v-1] = curlabel;
	curlabel--;
}

std::vector<int> topo_sort(std::vector<int> vertices, std::vector<int> edges){
	int num_ver = vertices.size();
	explored = std::vector<bool> (num_ver, false);
	curlabel = num_ver;
	for (int v : vertices){
		if (!explored[v-1]){
			dfs_topo(vertices, edges, v);
		}
	}
	return labels;
}
